[incubator/spring-cloud-data-flow] Update to new app version 1.3.0

trisberg opened the pull request:
- Add new spring-cloud-skipper component

- Add RBAC support

- Update requirements versions

*****************************************************************************************

Thank you for contributing to kubernetes/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/kubernetes/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the 
history. This will make it easier to identify new changes. The PR will be squashed 
anyways when it is merged. Thanks.

*****************************************************************************************

trisberg committed:
[incubator/spring-cloud-data-flow] Update to new app version 1.3.0

- Add new spring-cloud-skipper component

- Add RBAC support

- Update requirements versions

trisberg commented:
/assign @unguiculus

unguiculus commented:
Our RBAC best practices have been revised. See https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/rbac.md. Please update accordingly.

Here's an example for a chart that already implements this: https://github.com/kubernetes/charts/tree/master/stable/cert-manager

trisberg committed:
Updating chart for RBAC best practices

trisberg commented:
/assign @unguiculus

unguiculus commented on incubator/spring-cloud-data-flow/templates/_helpers.tpl:
```diff
@@ -32,3 +32,14 @@ We truncate at 63 chars because some Kubernetes name fields are limited to this
+{{/*
+Create the name of the service account to use
+*/}}
+{{- define "serviceAccountName" -}}
```
Should be `{{- define "scdf.serviceAccountName" -}}`

unguiculus commented on incubator/spring-cloud-data-flow/templates/server-rbac.yaml:
```diff
@@ -2,8 +2,13 @@
 apiVersion: rbac.authorization.k8s.io/{{ required "A valid .Values.rbac.apiVersion entry required!" .Values.rbac.apiVersion }}
 metadata:
-  name: {{ template "fullname" . }}-role
+  name: {{ template "fullname" . }}
```
I wonder why this resource explicitly specifies the namespace as opposed to the other resources in the chart. I'd suggest you remove it.

unguiculus reviewed (changes requested):
Please remove the `name` and `fullname` templates from `_helpers.tpl`. Only namespaced templates should be used.

trisberg committed:
Addressing further review comments

- use "scdf" prefix for name/fullname/serviceAccountName templates

- remove namespace from Role definition

- update to latest image tags for Data Flow server

trisberg commented:
/assign @unguiculus

unguiculus commented on incubator/spring-cloud-data-flow/values.yaml:
```diff
@@ -22,6 +35,24 @@ server:
+  service:
+    type: ClusterIP
+    externalPort: 80
+    internalPort: 7577
```
It probably doesn't make much sense to make container ports configurable. Changing them would break things unless the Spring config is updated as well.

trisberg replied:
Removing most of the port configs, only leaving the main server port as configurable

unguiculus commented on incubator/spring-cloud-data-flow/README.md:
```diff
@@ -36,6 +36,12 @@ Note that this chart pulls in many different Docker images so can take a while t
+
+| Parameter     | Description                 | Default                   |
+| ------------- | --------------------------- | ------------------------- |
+| rbac.create   | Create RBAC configurations  | false
```
Service account missing.

trisberg replied:
Adding

unguiculus commented on incubator/spring-cloud-data-flow/templates/skipper-deployment.yaml:
```diff
@@ -0,0 +1,60 @@
+              name: {{ printf "%s-%s" .Release.Name "rabbitmq" | trunc 63 | trimSuffix "-" }}
+              key: rabbitmq-password
+        - name: SPRING_APPLICATION_JSON
+          value: '{"spring.cloud.skipper.server.enableLocalPlatform" : false, "spring.cloud.skipper.server.platform.kubernetes.accounts.{{ .Values.skipper.platformName }}.environmentVariables" : "SPRING_RABBITMQ_HOST=${{ printf "{%s" .Release.Name | upper }}_RABBITMQ_SERVICE_HOST},SPRING_RABBITMQ_PORT=${{ printf "{%s" .Release.Name | upper }}_RABBITMQ_SERVICE_PORT_AMQP},SPRING_RABBITMQ_USERNAME={ ...
```
Wouldn't it be better to put that in a configmap?

trisberg replied:
For the long term, yes, that would be a better approach. The current version of the Skipper server doesn't use the `spring-cloud-starter-kubernetes-config` module so there is no built-in support for reading ConfigMaps.

unguiculus replied:
I meant that you could use a configmap as the source of the environment variable.

trisberg committed:
Address review comments

- add seviceAccount config to README

- remove most port config options

- move SPRING_APPLICATION_JSON to ConfigMap

trisberg committed:
Update Chart.yaml with latest release version

trisberg commented:
/assign @unguiculus

unguiculus commented:
Could you use named port and refer to them by name in probes and services?

unguiculus commented:
/ok-to-test

trisberg committed:
Adding names to container ports

- refer to ports by name for services and probes

- remove TCP protocol since it is the default

- add probes for skipper deployment

trisberg commented:
/retest

trisberg committed:
Avoid warning message for skipper init-mysql-database init container

trisberg commented:
/retest

trisberg commented:
Looks like an init container fails for the "skipper" component. I don't see any logs for the init containers so it's hard to tell why it is failing, works fine when I test on GKE.

trisberg committed:
Modifying init container init-wait-mysql for skipper component

trisberg commented:
/retest

sabbyanandan commented:
Hi, @unguiculus. Any advice how to troubleshoot the test failures in the CI box? 

Other than `W0306 21:39:14.519] spring--6150-data-flow-skipper-775dcf44fd-tt5gz   0/1       Init:Error   1         55s`, not much information in the [pull-charts-e2e/6150](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/charts/3525/pull-charts-e2e/6150/) logs. As an aside, we have verified the chart-PR on a few cluster variants on GKE and in minikube. 

Any help is appreciated.

unguiculus commented:
I'm getting the same error on a GKE cluster with RBAC disabled.

```
Name:           rousing-buffalo-data-flow-skipper-86998b9d79-v4wfk
Namespace:      default
Node:           gke-plasma-k8s-blue-default-pool-3e0f541e-z71l/10.150.0.2
Start Time:     Thu, 08 Mar 2018 17:33:03 +0100
Labels:         app=spring-cloud-data-flow
                component=skipper
                pod-template-hash=4255465835
                release=rousing-buffalo
Annotations:    kubernetes.io/limit-ranger=LimitRanger plugin set: cpu request for container rousing-buffalo-data-flow-skipper; cpu request for init container init-wait-mysql; cpu request for init container init-mysq...
Status:         Pending
IP:             10.151.0.119
Controlled By:  ReplicaSet/rousing-buffalo-data-flow-skipper-86998b9d79
Init Containers:
  init-wait-mysql:
    Container ID:  docker://988e76c12c5f8c21b22ca01beb54774980ace52d3121753e5b370f5d56736023
    Image:         busybox
    Image ID:      docker-pullable://busybox@sha256:c7b0a24019b0e6eda714ec0fa137ad42bc44a754d9cea17d14fba3a80ccc1ee4
    Port:          <none>
    Command:
      sh
      -c
      until nc -w3 -z rousing-buffalo-mysql 3306; do echo waiting for mysql; sleep 3; done;
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Thu, 08 Mar 2018 17:33:04 +0100
      Finished:     Thu, 08 Mar 2018 17:33:53 +0100
    Ready:          True
    Restart Count:  0
    Requests:
      cpu:        100m
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from rousing-buffalo-data-flow-token-4n25b (ro)
  init-mysql-database:
    Container ID:  docker://7aa182b0cb9033663169967376f8ba039dab925cf311ef160aa669c944f34e39
    Image:         mysql:5.7.14
    Image ID:      docker-pullable://mysql@sha256:c8f03238ca1783d25af320877f063a36dbfce0daa56a7b4955e6c6e05ab5c70b
    Port:          <none>
    Command:
      sh
      -c
      mysql -h ${ROUSING-BUFFALO_MYSQL_SERVICE_HOST} -u root -e "CREATE DATABASE IF NOT EXISTS skipper;"
    State:          Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Thu, 08 Mar 2018 17:34:56 +0100
      Finished:     Thu, 08 Mar 2018 17:34:56 +0100
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Thu, 08 Mar 2018 17:34:25 +0100
      Finished:     Thu, 08 Mar 2018 17:34:25 +0100
    Ready:          False
    Restart Count:  3
    Requests:
      cpu:  100m
    Environment:
      MYSQL_PWD:  <set to the key 'mysql-root-password' in secret 'rousing-buffalo-mysql'>  Optional: false
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from rousing-buffalo-data-flow-token-4n25b (ro)
Containers:
  rousing-buffalo-data-flow-skipper:
    Container ID:
    Image:          springcloud/spring-cloud-skipper-server:1.0.1.RELEASE
    Image ID:
    Port:           7577/TCP
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Requests:
      cpu:      100m
    Liveness:   http-get http://:http/api/about delay=90s timeout=5s period=15s #success=1 #failure=3
    Readiness:  http-get http://:http/api/about delay=30s timeout=5s period=15s #success=1 #failure=3
    Environment:
      SPRING_DATASOURCE_URL:                jdbc:mysql://${ROUSING-BUFFALO_MYSQL_SERVICE_HOST}:3306/skipper
      SPRING_DATASOURCE_USERNAME:           root
      SPRING_DATASOURCE_PASSWORD:           <set to the key 'mysql-root-password' in secret 'rousing-buffalo-mysql'>  Optional: false
      SPRING_DATASOURCE_DRIVER_CLASS_NAME:  org.mariadb.jdbc.Driver
      RABBITMQ_PASSWORD:                    <set to the key 'rabbitmq-password' in secret 'rousing-buffalo-rabbitmq'>                     Optional: false
      SPRING_APPLICATION_JSON:              <set to the key 'spring.application.json' of config map 'rousing-buffalo-data-flow-skipper'>  Optional: false
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from rousing-buffalo-data-flow-token-4n25b (ro)
Conditions:
  Type           Status
  Initialized    False
  Ready          False
  PodScheduled   True
Volumes:
  rousing-buffalo-data-flow-token-4n25b:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  rousing-buffalo-data-flow-token-4n25b
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason                 Age                From                                                     Message
  ----     ------                 ----               ----                                                     -------
  Normal   Scheduled              2m                 default-scheduler                                        Successfully assigned rousing-buffalo-data-flow-skipper-86998b9d79-v4wfk to gke-plasma-k8s-blue-default-pool-3e0f541e-z71l
  Warning  FailedScheduling       2m                 default-scheduler                                        AssumePod failed: pod default/rousing-buffalo-data-flow-skipper-86998b9d79-v4wfk state wasn't initial but get assumed
  Warning  FailedScheduling       2m                 default-scheduler                                        Binding rejected: Operation cannot be fulfilled on pods/binding "rousing-buffalo-data-flow-skipper-86998b9d79-v4wfk": pod rousing-buffalo-data-flow-skipper-86998b9d79-v4wfk is already assigned to node "gke-plasma-k8s-blue-default-pool-3e0f541e-z71l"
  Normal   SuccessfulMountVolume  2m                 kubelet, gke-plasma-k8s-blue-default-pool-3e0f541e-z71l  MountVolume.SetUp succeeded for volume "rousing-buffalo-data-flow-token-4n25b"
  Normal   Pulling                2m                 kubelet, gke-plasma-k8s-blue-default-pool-3e0f541e-z71l  pulling image "busybox"
  Normal   Pulled                 2m                 kubelet, gke-plasma-k8s-blue-default-pool-3e0f541e-z71l  Successfully pulled image "busybox"
  Normal   Created                2m                 kubelet, gke-plasma-k8s-blue-default-pool-3e0f541e-z71l  Created container
  Normal   Started                2m                 kubelet, gke-plasma-k8s-blue-default-pool-3e0f541e-z71l  Started container
  Normal   Pulling                1m                 kubelet, gke-plasma-k8s-blue-default-pool-3e0f541e-z71l  pulling image "mysql:5.7.14"
  Normal   Pulled                 58s                kubelet, gke-plasma-k8s-blue-default-pool-3e0f541e-z71l  Successfully pulled image "mysql:5.7.14"
  Normal   Created                10s (x4 over 58s)  kubelet, gke-plasma-k8s-blue-default-pool-3e0f541e-z71l  Created container
  Normal   Started                10s (x4 over 58s)  kubelet, gke-plasma-k8s-blue-default-pool-3e0f541e-z71l  Started container
  Normal   Pulled                 10s (x3 over 57s)  kubelet, gke-plasma-k8s-blue-default-pool-3e0f541e-z71l  Container image "mysql:5.7.14" already present on machine
  Warning  BackOff                9s (x4 over 56s)   kubelet, gke-plasma-k8s-blue-default-pool-3e0f541e-z71l  Back-off restarting failed container
```

init-wait-mysql
```
waiting for mysql
waiting for mysql
waiting for mysql
waiting for mysql
waiting for mysql
waiting for mysql
waiting for mysql
waiting for mysql 
```

init-mysql-database
```
ERROR 2005 (HY000): Unknown MySQL server host 'BUFFALO_MYSQL_SERVICE_HOST' (0)
```

trisberg committed:
Fix env var references that include the release name

- replaces any '-' characters with '_' to create valid env var references

trisberg commented:
/retest

sabbyanandan commented:
@unguiculus: Thanks for trying it out and the observations. Thomas had identified the culprit to be on unpacking of `-` vs. `_` character in the release prefix. It looks like, with his latest commit, the build/test plan is green.

unguiculus commented:
/lgtm

trisberg commented:
@unguiculus thanks for helping out with this. Would you know how long it usually takes for chart updates to show up at https://hub.kubeapps.com/ ?

unguiculus commented:
The sync to the repo should happen fairly quickly. I don't know how fast https://hub.kubeapps.com/ picks that up. @prydonius should now. But it seems there's currently a problem with the sync.

unguiculus commented:
@trisberg The sync is currently broken. A fix is being worked on. Sorry about that.

unguiculus commented:
@trisberg Sync is working again.
